File: 21-Memento/src/go/text_editor.go
package main

// TextEditor is the Originator — it snapshots and restores its own state.
type TextEditor struct {
	content string
}

func (e *TextEditor) Type(text string) {
	e.content += text
}

func (e *TextEditor) Content() string {
	return e.content
}

// Save snapshots the current state into a memento.
func (e *TextEditor) Save(label string) *EditorMemento {
	return &EditorMemento{content: e.content, label: label}
}

// Restore resets state from a previously created memento.
func (e *TextEditor) Restore(memento *EditorMemento) {
	e.content = memento.content
}

File: 21-Memento/src/rust/text_editor.rs
use crate::editor_memento::EditorMemento;

/// Originator — snapshots and restores its own private state.
pub struct TextEditor {
    content: String,
}

impl TextEditor {
    pub fn new() -> Self {
        TextEditor { content: String::new() }
    }

    // `type` is a Rust keyword, so the method is named `type_text`.
    pub fn type_text(&mut self, text: &str) {
        self.content.push_str(text);
    }

    pub fn content(&self) -> &str {
        &self.content
    }

    /// Snapshot the current state into a memento.
    pub fn save(&self, label: &str) -> EditorMemento {
        EditorMemento::new(self.content.clone(), label.to_string())
    }

    /// Restore state from a previously created memento.
    pub fn restore(&mut self, memento: &EditorMemento) {
        self.content = memento.content().to_string();
    }
}
